### ISSUE TITLE
[QUESTION] How to build an android LKM

### ISSUE THREAD

**Issue by LovableFatty** [ClangBuiltLinux/linux#1102]
Hi, I'm a college newbie so I still new to linux in general. My end goal is to create a LKM that can run on android so I can log system calls for a project.

I'm confused on how to use Clang to build a kernel. After I build this kernel, what is the next step towards cross compiling it so I can push it using adb push to my android emulator?

**Comment by LovableFatty** [comment]
My steps:
git clone https://android.googlesource.com/kernel/goldfish
…
cd goldfish
git checkout -t origin/android-goldfish-2.6.29 -b goldfish
Downloaded NDK r18b
make ARCH=arm goldfish_armv7_defconfig
make ARCH=arm menuconfig
    check "enable loadable module support"
    check "module unloading"
make ARCH=arm SUB_ARM=arm CROSS_COMPILE=$/home/noah/Downloads/android-ndk-r21d/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/bin/arm-linux-androideabi-
![image](https://user-images.githubusercontent.com/31664445/88244823-d36c7200-cc49-11ea-9153-f95f4f0fb383.png)
The first issue I see is that it states "arm-linux-androideabi-gcc: Not Found" But I see a arm-linux-androideabi-gcc in the bottom right of this pic (which is the path I put for the cross_compile)
![image](https://user-images.githubusercontent.com/31664445/88244862-f8f97b80-cc49-11ea-81b9-aed2379ac557.png)

**Comment by LovableFatty** [comment]
Thought it was an issue with my path but I changed it to $HOME, and still the same issue
![image](https://user-images.githubusercontent.com/31664445/88245474-319a5480-cc4c-11ea-861c-16b6156460c5.png)

**Comment by LovableFatty** [comment]
![r18b_clang_error](https://user-images.githubusercontent.com/31664445/88246782-6c05f080-cc50-11ea-95ad-3705b24b2e9a.png)
Getting an issue with Clang now as well as Error 1 & Error 2

**Comment by LovableFatty** [comment]
> Hi Noah,
> 
> Few things initially:
> 
> 1. If you ran `git checkout -t origin/android-goldfish-2.6.29 -b goldfish`, you cloned a kernel that has not seen a commit in six years from Google and no upstream commits for probably ten years. You should use a newer kernel such as 4.14, which appears to be the most recent goldfish kernel.
> 2. Your `make` command seems to be missing `CC=clang`, which is how you actually build with Clang.
> 
> I had [a conversation with someone else](https://github.com/nathanchance/android-kernel-clang/issues/11) on my [android-kernel-clang](https://github.com/nathanchance/android-kernel-clang) repo along these lines before. I would recommend looking into Cuttlefish if you are looking to kernel hack with a virtual device. The goldfish kernel is rather hacked up from my understanding whereas Cuttlefish can technically boot mainline kernels (although you should probably use [kernel/common](https://android.googlesource.com/kernel/common) to start).
> 
> I wrote [a rather long blog post](https://nathanchance.dev/posts/building-using-cuttlefish/) on how to build a kernel for Cuttlefish that I think is still accurate to this day.
> 
> There is a decent amount of documentation about it on Google's website as well:
> 
> https://source.android.com/setup/create/cuttlefish
> 
> https://android.googlesource.com/device/google/cuttlefish/

Hi Nathan, thanks for your help. I've followed your cuttlefish guide but I ran into a problem.
My current steps:
$ sudo apt install -y build-essential devscripts fakeroot git psmisc qemu-kvm unzip zip
$ mkdir -p /home/Documents/noah/cuttlefish-playground/src
$ git clone https://github.com/google/android-cuttlefish /home/Documents/noah/cuttlefish-playground/src/android-cuttlefish
$ cd /home/Documents/noah/cuttlefish-playground/src/android-cuttlefish
$ sudo apt install -y cdbs config-package-dev debhelper
$ debuild -i -us -uc -b
$ sudo dpkg -i ../cuttlefish-common_*_amd64.deb
$ sudo apt-get install -f -y
$ rm ../cuttlefish-*
$ sudo usermod -aG cvdnetwork "${USER}"
$ sudo usermod -aG kvm "${USER}"
$ mkdir -p  /home/Documents/noah/cuttlefish-playground/cuttlefish-fs
$ cd /home/Documents/noah/cuttlefish-playground/cuttlefish-fs
$ sudo curl -LSs https://ci.android.com/builds/submitted/6173731/aosp_cf_x86_phone-userdebug/latest/cvd-host_package.tar.gz
$ tar -zxvf "aosp file" 
HOME=${PWD} ./bin/launch_cvd -daemon | ERROR OCCURS HERE
![image](https://user-images.githubusercontent.com/31664445/88336520-fea69e00-cce9-11ea-9004-a84f6d0ab0cb.png)

**Comment by LovableFatty** [comment]
> Hi Noah,
> 
> Few things initially:
> 
> 1. If you ran `git checkout -t origin/android-goldfish-2.6.29 -b goldfish`, you cloned a kernel that has not seen a commit in six years from Google and no upstream commits for probably ten years. You should use a newer kernel such as 4.14, which appears to be the most recent goldfish kernel.
> 2. Your `make` command seems to be missing `CC=clang`, which is how you actually build with Clang.
> 
> I had [a conversation with someone else](https://github.com/nathanchance/android-kernel-clang/issues/11) on my [android-kernel-clang](https://github.com/nathanchance/android-kernel-clang) repo along these lines before. I would recommend looking into Cuttlefish if you are looking to kernel hack with a virtual device. The goldfish kernel is rather hacked up from my understanding whereas Cuttlefish can technically boot mainline kernels (although you should probably use [kernel/common](https://android.googlesource.com/kernel/common) to start).
> 
> I wrote [a rather long blog post](https://nathanchance.dev/posts/building-using-cuttlefish/) on how to build a kernel for Cuttlefish that I think is still accurate to this day.
> 
> There is a decent amount of documentation about it on Google's website as well:
> 
> https://source.android.com/setup/create/cuttlefish
> 
> https://android.googlesource.com/device/google/cuttlefish/
Hi Nate. I received a error during your Cuttlefish build guide.
My steps:
$ sudo apt install -y build-essential devscripts fakeroot git psmisc qemu-kvm unzip zip
$ mkdir -p /home/Documents/noah/cuttlefish-playground/src
$ git clone https://github.com/google/android-cuttlefish /home/Documents/noah/cuttlefish-playground/src/android-cuttlefish
$ cd /home/Documents/noah/cuttlefish-playground/src/android-cuttlefish
$ sudo apt install -y cdbs config-package-dev debhelper
$ debuild -i -us -uc -b
$ sudo dpkg -i ../cuttlefish-common_*_amd64.deb
$ sudo apt-get install -f -y
$ rm ../cuttlefish-*
$ sudo usermod -aG cvdnetwork "${USER}"
$ sudo usermod -aG kvm "${USER}"
$ mkdir -p  /home/Documents/noah/cuttlefish-playground/cuttlefish-fs
$ cd /home/Documents/noah/cuttlefish-playground/cuttlefish-fs
Downloaded: https://ci.android.com/builds/branches/aosp-master/grid? | aosp_cf_x86_phone-userdebug @ 6703980
From this link I downloaded aosp_cf_x86_phone-img-6703980.zip & cvd-host_package.tar in /cuttlefish-fs
$ HOME=${PWD} ./bin/launch_cvd -daemon
![image](https://user-images.githubusercontent.com/31664445/88341768-0d458300-ccf3-11ea-9cfa-b7a0417621f8.png)

**Comment by dileks** [comment]
> I wrote [a rather long blog post](https://nathanchance.dev/posts/building-using-cuttlefish/) on how to build a kernel for Cuttlefish that I think is still accurate to this day.

Can you add this to [1]?
Thanks.

[1] https://github.com/ClangBuiltLinux/linux/wiki/Talks,-Presentations,-and-Communications

**Comment by LovableFatty** [comment]
> Hi Noah,
> 
> Few things initially:
> 
> 1. If you ran `git checkout -t origin/android-goldfish-2.6.29 -b goldfish`, you cloned a kernel that has not seen a commit in six years from Google and no upstream commits for probably ten years. You should use a newer kernel such as 4.14, which appears to be the most recent goldfish kernel.
> 2. Your `make` command seems to be missing `CC=clang`, which is how you actually build with Clang.
> 
> I had [a conversation with someone else](https://github.com/nathanchance/android-kernel-clang/issues/11) on my [android-kernel-clang](https://github.com/nathanchance/android-kernel-clang) repo along these lines before. I would recommend looking into Cuttlefish if you are looking to kernel hack with a virtual device. The goldfish kernel is rather hacked up from my understanding whereas Cuttlefish can technically boot mainline kernels (although you should probably use [kernel/common](https://android.googlesource.com/kernel/common) to start).
> 
> I wrote [a rather long blog post](https://nathanchance.dev/posts/building-using-cuttlefish/) on how to build a kernel for Cuttlefish that I think is still accurate to this day.
> 
> There is a decent amount of documentation about it on Google's website as well:
> 
> https://source.android.com/setup/create/cuttlefish
> 
> https://android.googlesource.com/device/google/cuttlefish/



Hi Nathan, so I've been trying out cuttlefish and I've encountered an issue . 
$ sudo apt install -y build-essential devscripts fakeroot git psmisc qemu-kvm unzip zip
$ mkdir -p /home/Documents/noah/cuttlefish-playground/src
$ git clone https://github.com/google/android-cuttlefish /home/Documents/noah/cuttlefish-playground/src/android-cuttlefish
$ cd /home/Documents/noah/cuttlefish-playground/src/android-cuttlefish
$ sudo apt install -y cdbs config-package-dev debhelper
$ debuild -i -us -uc -b
// note a long list of things just came out
$ sudo dpkg -i ../cuttlefish-common_*_amd64.deb
// try sudo apt-get install -f -y if the step above fails and try it again
$ rm ../cuttlefish-*
$ sudo usermod -aG cvdnetwork "${USER}"
$ sudo usermod -aG kvm "${USER}"
$ mkdir /home/Documents/noah/cuttlefish-playground/cuttlefish-fs
$ cd /home/Documents/noah/cuttlefish-playground/cuttlefish-fs
// Go to https://ci.android.com/builds/submitted/6173731/aosp_cf_x86_phone-userdebug/latest and 
download cvd-host_package.tar.gz and aosp_cf_x86_phone-img-xxxx.zip
$ unzip *.zip 
$ tar -xvf cvd-host_package.tar.gz

$ HOME=${PWD} ./bin/launch_cvd -daemon | ERROR 

$ sudo apt install -y build-essential devscripts fakeroot git psmisc qemu-kvm unzip zip
$ mkdir -p /home/Documents/noah/cuttlefish-playground/src
$ git clone https://github.com/google/android-cuttlefish /home/Documents/noah/cuttlefish-playground/src/android-cuttlefish
$ cd /home/Documents/noah/cuttlefish-playground/src/android-cuttlefish
$ sudo apt install -y cdbs config-package-dev debhelper
$ debuild -i -us -uc -b
// note a long list of things just came out
$ sudo dpkg -i ../cuttlefish-common_*_amd64.deb
// try sudo apt-get install -f -y if the step above fails and try it again
$ rm ../cuttlefish-*
$ sudo usermod -aG cvdnetwork "${USER}"
$ sudo usermod -aG kvm "${USER}"
$ mkdir /home/Documents/noah/cuttlefish-playground/cuttlefish-fs
$ cd /home/Documents/noah/cuttlefish-playground/cuttlefish-fs
// Go to https://ci.android.com/builds/submitted/6173731/aosp_cf_x86_phone-userdebug/latest and 
download cvd-host_package.tar.gz and aosp_cf_x86_phone-img-xxxx.zip
$ unzip *.zip 
$ tar -xvf cvd-host_package.tar.gz

$ HOME=${PWD} ./bin/launch_cvd -daemon | ERROR 
$ sudo apt install -y build-essential devscripts fakeroot git psmisc qemu-kvm unzip zip
$ mkdir -p /home/Documents/noah/cuttlefish-playground/src
$ git clone https://github.com/google/android-cuttlefish /home/Documents/noah/cuttlefish-playground/src/android-cuttlefish
$ cd /home/Documents/noah/cuttlefish-playground/src/android-cuttlefish
$ sudo apt install -y cdbs config-package-dev debhelper
$ debuild -i -us -uc -b
// note a long list of things just came out
$ sudo dpkg -i ../cuttlefish-common_*_amd64.deb
// try sudo apt-get install -f -y if the step above fails and try it again
$ rm ../cuttlefish-*
$ sudo usermod -aG cvdnetwork "${USER}"
$ sudo usermod -aG kvm "${USER}"
$ mkdir /home/Documents/noah/cuttlefish-playground/cuttlefish-fs
$ cd /home/Documents/noah/cuttlefish-playground/cuttlefish-fs
// Go to https://ci.android.com/builds/submitted/6173731/aosp_cf_x86_phone-userdebug/latest and 
download cvd-host_package.tar.gz and aosp_cf_x86_phone-img-xxxx.zip
$ unzip *.zip 
$ tar -xvf cvd-host_package.tar.gz

$ HOME=${PWD} ./bin/launch_cvd -daemon | ERROR 

$ sudo apt install -y build-essential devscripts fakeroot git psmisc qemu-kvm unzip zip
$ mkdir -p /home/Documents/noah/cuttlefish-playground/src
$ git clone https://github.com/google/android-cuttlefish /home/Documents/noah/cuttlefish-playground/src/android-cuttlefish
$ cd /home/Documents/noah/cuttlefish-playground/src/android-cuttlefish
$ sudo apt install -y cdbs config-package-dev debhelper
$ debuild -i -us -uc -b
// note a long list of things just came out
$ sudo dpkg -i ../cuttlefish-common_*_amd64.deb
// try sudo apt-get install -f -y if the step above fails and try it again
$ rm ../cuttlefish-*
$ sudo usermod -aG cvdnetwork "${USER}"
$ sudo usermod -aG kvm "${USER}"
$ mkdir /home/Documents/noah/cuttlefish-playground/cuttlefish-fs
$ cd /home/Documents/noah/cuttlefish-playground/cuttlefish-fs
// Go to https://ci.android.com/builds/submitted/6173731/aosp_cf_x86_phone-userdebug/latest and 
download cvd-host_package.tar.gz and aosp_cf_x86_phone-img-xxxx.zip
$ unzip *.zip 
$ tar -xvf cvd-host_package.tar.gz

$ HOME=${PWD} ./bin/launch_cvd -daemon | ERROR 

------------ ERROR ---------------
noah@kali-noah:~/cuttlefish-playground/cuttlefish-fs$ HOME=${PWD} ./bin/launch_cvd -daemon
launch_cvd I 07-29 15:28:07 16888 16888 subprocess.cpp:286] Started (pid: 16915): /home/noah/cuttlefish-playground/cuttlefish-fs/bin/assemble_cvd
launch_cvd I 07-29 15:28:07 16888 16888 subprocess.cpp:289] --daemon=true
launch_cvd I 07-29 15:28:07 16888 16888 subprocess.cpp:286] Started (pid: 16916): /home/noah/cuttlefish-playground/cuttlefish-fs/bin/run_cvd
assemble_cvd W 07-29 15:28:07 16915 16915 cuttlefish_config.cpp:53] No user or non-vsoc user, returning default config
assemble_cvd I 07-29 15:28:07 16915 16915 flags.cc:468] Assuming prior files of /home/noah/cuttlefish-playground/cuttlefish-fs/cuttlefish_runtime/* /home/noah/cuttlefish-playground/cuttlefish-fs/cuttlefish_assembly/* /home/noah/cuttlefish-playground/cuttlefish-fs/.cuttlefish.sh /home/noah/cuttlefish-playground/cuttlefish-fs/.cuttlefish_config.json
assemble_cvd I 07-29 15:28:07 16915 16915 flags.cc:641] Setting up /home/noah/cuttlefish-playground/cuttlefish-fs/cuttlefish_assembly
assemble_cvd I 07-29 15:28:07 16915 16915 flags.cc:650] Setting up /home/noah/cuttlefish-playground/cuttlefish-fs/cuttlefish_runtime
assemble_cvd E 07-29 15:28:07 16915 16915 fetcher_config.cpp:212] Could not find file ending in kernel
assemble_cvd E 07-29 15:28:07 16915 16915 fetcher_config.cpp:212] Could not find file ending in initramfs.img
assemble_cvd E 07-29 15:28:07 16915 16915 fetcher_config.cpp:212] Could not find file ending in kernel
assemble_cvd E 07-29 15:28:07 16915 16915 fetcher_config.cpp:212] Could not find file ending in initramfs.img
assemble_cvd I 07-29 15:28:07 16915 16915 data_image.cc:148] misc partition image: creating empty
assemble_cvd I 07-29 15:28:07 16915 16915 data_image.cc:67] Creating /home/noah/cuttlefish-playground/cuttlefish-fs//misc.img
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:286] Started (pid: 16929): /bin/dd
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:289] if=/dev/zero
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:289] of=/home/noah/cuttlefish-playground/cuttlefish-fs//misc.img
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:289] bs=1M
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:289] count=1
1+0 records in
1+0 records out
1048576 bytes (1.0 MB, 1.0 MiB) copied, 0.000540867 s, 1.9 GB/s
assemble_cvd I 07-29 15:28:07 16915 16915 data_image.cc:134] /home/noah/cuttlefish-playground/cuttlefish-fs//userdata.img exists. Not creating it.
assemble_cvd I 07-29 15:28:07 16915 16915 data_image.cc:67] Creating /home/noah/cuttlefish-playground/cuttlefish-fs//metadata.img
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:286] Started (pid: 16930): /bin/dd
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:289] if=/dev/zero
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:289] of=/home/noah/cuttlefish-playground/cuttlefish-fs//metadata.img
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:289] bs=1M
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:289] count=16
16+0 records in
16+0 records out
16777216 bytes (17 MB, 16 MiB) copied, 0.005725 s, 2.9 GB/s
assemble_cvd I 07-29 15:28:07 16915 16915 data_image.cc:67] Creating /home/noah/cuttlefish-playground/cuttlefish-fs/cuttlefish_runtime/access-kregistry
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:286] Started (pid: 16931): /bin/dd
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:289] if=/dev/zero
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:289] of=/home/noah/cuttlefish-playground/cuttlefish-fs/cuttlefish_runtime/access-kregistry
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:289] bs=64K
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:289] count=1
1+0 records in
1+0 records out
65536 bytes (66 kB, 64 KiB) copied, 6.2466e-05 s, 1.0 GB/s
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:286] Started (pid: 16932): /home/noah/cuttlefish-playground/cuttlefish-fs/bin/cf_bpttool
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:289] make_table
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:289] --input=/dev/stdin
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:289] --output_json=/dev/stdout
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:286] Started (pid: 16934): /home/noah/cuttlefish-playground/cuttlefish-fs/bin/cf_bpttool
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:289] make_table
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:289] --input=/dev/stdin
assemble_cvd I 07-29 15:28:07 16915 16915 subprocess.cpp:289] --output_gpt=/dev/stdout
launch_cvd I 07-29 15:28:07 16888 16888 launch_cvd.cc:124] assemble_cvd exited successfully.
run_cvd W 07-29 15:28:07 16916 16916 cuttlefish_config.cpp:53] No user or non-vsoc user, returning default config
run_cvd I 07-29 15:28:07 16916 16916 subprocess.cpp:286] Started (pid: 16939): /bin/bash
run_cvd I 07-29 15:28:07 16916 16916 subprocess.cpp:289] -c
run_cvd I 07-29 15:28:07 16916 16916 subprocess.cpp:289] egrep -h -e "^iff:.*" /proc/*/fdinfo/*
run_cvd E 07-29 15:28:07 16916 16916 users.cpp:48] Group virtaccess does not exist
run_cvd I 07-29 15:28:07 16916 16916 main.cc:367] The following files contain useful debugging information:
run_cvd I 07-29 15:28:07 16916 16916 main.cc:369]   Launcher log: /home/noah/cuttlefish-playground/cuttlefish-fs/cuttlefish_runtime/launcher.log
run_cvd I 07-29 15:28:07 16916 16916 main.cc:371]   Android's logcat output: /home/noah/cuttlefish-playground/cuttlefish-fs/cuttlefish_runtime/logcat
run_cvd I 07-29 15:28:07 16916 16916 main.cc:372]   Kernel log: /home/noah/cuttlefish-playground/cuttlefish-fs/cuttlefish_runtime/kernel.log
run_cvd I 07-29 15:28:07 16916 16916 main.cc:373]   Instance configuration: /home/noah/cuttlefish-playground/cuttlefish-fs/cuttlefish_runtime/cuttlefish_config.json
run_cvd I 07-29 15:28:07 16916 16916 main.cc:374]   Instance environment: /home/noah/cuttlefish-playground/cuttlefish-fs/.cuttlefish.sh
run_cvd I 07-29 15:28:07 16916 16916 main.cc:375] To access the console run: socat file:$(tty),raw,echo=0 /home/noah/cuttlefish-playground/cuttlefish-fs/cuttlefish_runtime/console

------------------------------------
Been stuck there for 20+ minutes, I have a pretty good laptop running a 8th gen i7 so I don't think it should take that long.

**Comment by nathanchance** [comment]
Have you tried rebooting your computer? Additionally, is virtualization enabled in your computer's BIOS settings?

**Comment by LovableFatty** [comment]
> Have you tried rebooting your computer? Additionally, is virtualization enabled in your computer's BIOS settings?

Hey Nathan,

I was able to get a majority of things working, I'm at nearly the last step. I downloaded my kernel source as Common Android 5.4
Ran this command $ BUILD_CONFIG=cuttlefish-modules/build.config.cuttlefish.x86_64 ./build/build.sh
and I checked for a bzimage in out/android12-5.4/common/arch/x86/boot/bzImage
but I couldn't find a bzimage instead I found a kexec-bzimage64.c
can i use this as a kernel to run my emulator? 
![image](https://user-images.githubusercontent.com/31664445/89089099-92164980-d350-11ea-8279-242c427e70ef.png)
I also found a vmlinux which appears to be a kernel, does this one work instead of bzimage?
![image](https://user-images.githubusercontent.com/31664445/89089126-af4b1800-d350-11ea-8543-b5acaca42e8a.png)
